README.md
# IdentityCache

IdentityCache is a drop-in array replacement object cache designed to work with 
ORM and Identity Map patterns, providing Weakref-based unique 
identity-to-object resolution and smart caching of unused objects based on 
objects popularity  

## Abstract

The [Identity Map](https://en.wikipedia.org/wiki/Identity_map_pattern) pattern
is very common in ORM implementations. It consists of an intermediate cache
between the persistence and application layer, which is able to retrieve
object instances by an unique object id. It must satisfy two requirements:

* For as long as an Identity is in use (the application layer holds any 
  references to it), the Identity Map will retrieve the
  same instance when its id is queried. Otherwise, data consistency will be 
  compromised when the application layer mutates different instances of the same
  identity.
* For Identities not in use, the Cache provides performance benefits by keeping
  instances for eventual reuse. It should however have a mechanism to discard 
  cached instances so as not to consume excessive resources (i.e. memory).

Many implementations use a PHP native array to store an identity => instance
map; however this has the disadvantage of storing instances indefinitely,
ultimately filling up memory even though the references are not still in use.   
  
## Implementation

The IdentityCache implements Weakref-based caching to satisfy the requirements
and a popularity-based garbage collection algorithm to keep the size of the
cached unused instances into configurable limits. 

The `IdentityCache` and `IdentityMap` are designed as drop-in replacements for a
native PHP array. The `IdentityMap` does not provide caching functionality; as
soon as an object is not used anymore (its refcount drops to 0), the object is
freed.     

## Requirements

* For PHP > 7.4, use `Exteon\IdentityCache\Weakreference\IdentityMap`, 
  `Exteon\IdentityCache\Weakreference\IdentityMap` which use 
  [Weakreference](https://www.php.net/manual/en/class.weakreference.php)
* For PHP < 7.4, the [Weakref](https://pecl.php.net/package/Weakref) extension 
  is needed. Then use `Exteon\IdentityCache\Weakref\IdentityMap`, 
  `Exteon\IdentityCache\Weakref\IdentityMap`
* PHP < 7.2 not supported
 
## Usage

### Installing with `composer`

```shell script
composer config repositories.exteon-identity-map vcs https://github.com/exteon/identity-cache
composer require exteon/identity-cache
```

### IdentityMap

```php
$map = new \Exteon\IdentityCache\WeakReference\IdentityMap();

$instance = new stdClass();
$map[1] = $instance;

// While object is in use via $instance, the map will provide a reference to it
// via its id.

assert($map[1] === $instance);

// Put object out of use by releasing its reference

unset($instance);

// Object has been freed and its reference unset, because all references to it
// were released

assert(isset($map[1]) === false);
```

### IdentityCache

```php
$cache = new \Exteon\IdentityCache\WeakReference\IdentityCache([
    'trigger' => 'maxRetainedObjects',
    'maxRetainedObjects' => 1,
    'purgeStrategy' => 'popularity',
    'purgePressure' => 50
]);

// Create 2 object instances

$instance1 = new stdClass();
$instance2 = new stdClass();

$cache[1] = $instance1;
$cache[2] = $instance2;

// While objects are in use via $instance1, $instance2, the map holds references
// to them via their ids

assert($cache[1] === $instance1);
assert($cache[1] === $instance1);

// Increase instance 1's popularity by accessing it repeatedly

$cache[1];
$cache[1];
$cache[1];

// Put objects out of use by releasing their references

unset($instance1);
unset($instance2);

// maxRetainedObjects was configured to 1; instances will be purged

$cache->gc();

// instance 1's popularity was greater so it was preserved in the cache

assert(is_object($cache[1]));

// instance 2 was of a lesser popularity so it was purged

assert(isset($cache[2]) === false);
```

#### Configuration

An associative array can be passed to the constructor of 
`\Exteon\IdentityCache\WeakReference\IdentityCache`. Example usage, default and 
possible values are shown below. If a configuration option is not passed to the 
constructor, its default value will be used:

```php
use \Exteon\IdentityCache\WeakReference\IdentityCache;

$cache = new IdentityCache([
    //  Specifies the condition that triggers the cache to purge unused 
    //  references.
    //  Possible values:
    //      'maxRetainedObjects'    :   purging will be initiated when we are
    //                                  caching a number of references greater
    //                                  than the maxRetainedObjects parameter
    //      'maxScriptMemory'       :   purging will be initiated when the total
    //                                  memory consumed by the running script is
    //                                  greater than the value specified in the
    //                                  maxScriptMemory parameter.
    //      'none'                  :   purging will not be done. This can be
    //                                  changed later by using the setConfig()
    //                                  method.
    'trigger' => 'maxRetainedObjects',

    //  Number of objects that can be retained before purging if trigger is
    //  set to 'maxRetainedObjects'
    'maxRetainedObjects' => 1000,

    //  Maximum total memory consumed by the running script before purging if
    //  trigger is 'maxScriptMemory'. Format is the same as php.ini's sizes// 
    //  format.
    'maxScriptMemory' => '64M',

    //  When purging, which percent of the cached objects to purge?
    'purgePressure' => 10,

    //  When purging, how to select which instances to keep?
    //  Possible values:
    //      'popularity'    :   Keep the most popular objects. See the following 
    //                          section for more details on how popularity-based
    //                          caching works.
    //      'random'        :   Randomly purge the number of objects dictated
    //                          by purgePressure.  
    'purgeStrategy' => 'popularity',

    //  See the Popularity-based caching section for an explanation of this// 
    //  parameter
    'popularityDecay' => IdentityCache::getPopularityDecay(1000, 10000, 2)
]);
```

##### Popularity-based caching

When the `purgeStrategy` config option is `popularity`, a popularity index will 
be kept for the cached objects. An object's popularity increases by 1 whenever 
it is accessed from the cache.

In order to not keep forever objects that were once very popular but have not
been used in a long time, an object's popularity also decays exponentially every
time another object is accessed from the cache.

The value of the decay parameter can be specified by the `popularityDecay`
config option. This is a very sensitive float value that can be computed using 
the `getPopularityDecay()` method, as such:

```php
$popularityDecay = \Exteon\IdentityCache\WeakReference\IdentityCache::getPopularityDecay(
    1000,   //  initialPopularity
    10000,  //  rounds
    2       //  targetPopularity
);
```

The meaning of the parameters above, shown with default values, is:
If an object has at the present moment a popularity of `initialPopularity`
(1000), then other objects are accessed from the cache a number of `rounds` 
(10000) times, the popularity of the initial object will drop to 
`targetPopularity` (2). By tweaking the value of `popularityDecay`, you can
balance between keeping in the cache more of formerly popular objects or more of
recent objects.   

#### Acquiring objects

Mapped objects can be explicitly 'acquired', meaning the map/cache will hold 
them indefinitely even if they are no longer in use (they won't be purged). This 
is needed for 'dirty' objects that have been mutated and will need to be 
persisted later in a write-behind strategy scenario. 

Example:
```php
$cache = new \Exteon\IdentityCache\WeakReference\IdentityCache([
    'trigger' => 'maxRetainedObjects',
    'maxRetainedObjects' => 0
]);

$instance = new stdClass();
$cache[1] = $instance;

$instance->foo = 'bar';
$cache->acquire(1);

//  The acquired identity with id 1 will never be purged, even if the cache's
//  purge strategy is triggered by number of objects or memory trigger 

unset($instance);
$cache->gc();
assert(
    is_object($cache[1]) &&
    $cache[1]->foo === 'bar'
);

//  Acquired identities can be released and then they can be purged if no longer 
//  in use:

$cache->release(1);
$cache->gc();
assert(isset($cache[1]) === false);
```

#### Invoking the garbage collector explicitly

The garbage collector is normally invoked every time an identity is added to the
cache; it will then check if the trigger condition is satisfied and if so it
will proceed to purge excess objects from the cache. The gc can be also invoked
manually by calling the `gc()` method on `IdentityCache`.

### Native PHP array replacements

If the `Weakref` extension is not available, or for testing purposes, or for
an emergency in case the Weakref / WeakReference implementation is failing, we 
provide regular array-based replacements for the Weakref `IdentityMap` and 
`IdentityCache`:

`\Exteon\IdentityCache\NativeArray\IdentityMap`  
`\Exteon\IdentityCache\NativeArray\IdentityCache`

These implement the same interfaces as their WeakRef / WeakReference 
counterparts, so they can be dropped in their place, but they will not implement 
any reference releasing functionality: objects added will remain in the 
`IdentityMap` or `IdentityCache` until explicitly `unset()`.

CHANGELOG.md
## 1.2.0

#### New features

* Added `Factory` class to provide best available version for objects

## 1.1.0

#### New features

* Added `IIdentityMap::getKeys()` method for PHP8 compatibility
* Added `::canUse()` on `IIdentityMap` for greater generalisation
* Added PHP7.4/8 `WeakReference` map/cache in 
  `Exteon\IdentityCache\WeakReference` namespace

src/Strategy/CachingStrategy.php
<?php
    /**
     * @package Exteon\IdentityCache
     * @author Dinu Marina
     * @link https://github.com/exteon/identity-cache
     * @license https://www.apache.org/licenses/LICENSE-2.0
     */

    namespace Exteon\IdentityCache\Strategy;

    use Exception;
    use Exteon\MemoryHelper;
    use InvalidArgumentException;

    /**
     * The WeakRef IdentityCache implements, in addition to {@see IdentityMap},
     * caching of unused instances, to optimize performance for subsequent
     * fetches of the same identity. The cache size can be limited via the
     * configuration options, and the caching strategy can include
     * popularity-based caching.
     *
     * @package Exteon\IdentityCache\WeakRef
     */
    trait CachingStrategy {

        /** @var array */
        private $collectionGcHold = [];

        /** @var array */
        private $config = [
            'trigger' => 'maxRetainedObjects',
            'maxRetainedObjects' => 1000,
            'maxScriptMemory' => '64M',
            'purgePressure' => 10,
            'purgeStrategy' => 'popularity',
        ];

        /** @var int */
        private $trigger;

        /** @var int */
        private $purgeStrategy;

        /**
         * For the garbage collector, this collection holds the objects
         * popularity (number of instantiations from the DB) for every object.
         * Object ids are keys.
         *
         * @var int[]
         */
        private $objectsPopularity = [];

        /** @var int[] */
        private $objectsPopularityDecayUpdated = [];

        /** @var int */
        private $objectsPopularityDecayCounter = 0;

        /**
         * IdentityCache constructor.
         *
         * @param array $config
         * The config parameter can have the following keys/values. If a setting
         * is not specified, its default will be used that is listed below.
         * <br />
         * <ul>
         *      <li>'trigger' => 'maxRetainedObjects'<br>
         *          Specifies the condition that triggers the cache to purge
         *          unused references.<br>
         *          <br>
         *          Possible values:<ul>
         *              <li>'maxRetainedObjects' :<br>
         *                  purging will be initiated when we are caching a
         *                  number of references greater than the
         *                  maxRetainedObjects parameter<br><br></li>
         *              <li>'maxScriptMemory' :<br>
         *                  purging will be initiated when the total memory
         *                  consumed by the running script is greater than the
         *                  value specified in the maxScriptMemory parameter.
         *                  <br><br></li>
         *              <li>'none' :<br>
         *                  purging will not be done. This can be changed later
         *                  by using the {@see setConfig()} method.
         *                  <br></li></ul></li>
         *      <li>'maxRetainedObjects' => 1000<br>
         *          Number of objects that can be retained before purging if
         *          trigger is set to 'maxRetainedObjects'<br><br></li>
         *      <li>'maxScriptMemory' => '64M'<br>
         *          Maximum total memory consumed by the running script before
         *          purging if trigger is 'maxScriptMemory'. Format is the same
         *          as php.ini's sizes// format.<br><br></li>
         *      <li>'purgePressure' => 10<br>
         *          When purging, which percent of the cached objects to purge?
         *          <br><br></li>
         *      <li>'purgeStrategy' => 'popularity'<br>
         *          When purging, how to select which instances to keep?<br>
         *          <br>
         *          Possible values:<ul>
         *              <li>'popularity' :<br>
         *                  Keep the most popular objects. See the following
         *                  section for more details on how popularity-based
         *                  caching works.<br><br></li>
         *              <li>'random' :<br>
         *                  Randomly purge the number of objects dictated by
         *                  purgePressure.<br></li></ul></li>
         *      <li>'popularityDecay' => self::getPopularityDecay(1000, 10000, 2)
         *          <br>
         *          See {@see getPopularityDecay()} for an explanation of this
         *          value
         *          <br><br></li></ul>
         *
         * @throws Exception
         */
        public function __construct(array $config = []) {
            $this->setConfig($config);
        }

        /**
         * @param bool|float|int|string $offset
         * @return mixed|null
         * @throws Exception
         */
        public function offsetGet($offset) {
            $result = parent::offsetGet($offset);
            if ($result !== null) {
                if (
                    !array_key_exists(
                        $offset,
                        $this->collectionGcHold
                    ) &&
                    !$this->checkGcTrigger()
                ) {
                    $this->collectionGcHold[$offset] = $result;
                }
                if ($this->purgeStrategy === Constants::PURGE_STRATEGY_POPULARITY) {
                    $this->increasePopularity($offset);
                }
            }

            return $result;
        }

        /**
         * @param bool|float|int|string $offset
         * @param mixed $value
         * @throws Exception
         */
        public function offsetSet($offset, $value): void {
            parent::offsetSet($offset, $value);
            $this->collectionGcHold[$offset] = $value;
            if ($this->purgeStrategy === Constants::PURGE_STRATEGY_POPULARITY) {
                $this->increasePopularity($offset);
                $this->increasePopularity($offset);
            }
            $this->gc();
        }

        /**
         * @param bool|float|int|string $offset
         */
        public function offsetUnset($offset): void {
            parent::offsetUnset($offset);
            $this->releaseHold($offset);
            $this->release($offset);
            unset($this->objectsPopularity[$offset]);
            unset($this->objectsPopularityDecayUpdated[$offset]);
        }

        /**
         * @param bool|float|int|string $offset
         */
        protected function releaseHold($offset): void {
            unset($this->collectionGcHold[$offset]);
        }

        /**
         * @return bool
         * @throws Exception
         */
        protected function checkGcTrigger(): bool {
            switch ($this->trigger) {
                case Constants::TRIGGER_RETAINED_OBJECTS:
                    return (
                        count($this->collectionGcHold)
                        >
                        $this->config['maxRetainedObjects']
                    );
                case Constants::TRIGGER_SCRIPT_MEMORY:
                    return (
                        memory_get_usage()
                        >
                        MemoryHelper::bytesFromPhpIniString(
                            $this->config['maxScriptMemory']
                        )
                    );
                case Constants::TRIGGER_NONE:
                    return false;
                default:
                    throw new Exception('Invalid trigger value');
            }
        }

        /**
         * @throws Exception
         */
        public function gc(): void {
            if ($this->checkGcTrigger()) {
                $this->purge();
            }
        }

        protected function purge(): void {
            switch ($this->purgeStrategy) {
                case Constants::PURGE_STRATEGY_POPULARITY:
                    foreach ($this->collectionGcHold as $offset => $value) {
                        $this->decay($offset);
                    }
                    uksort(
                        $this->collectionGcHold,
                        function ($offset1, $offset2) {
                            $pop1 = $this->objectsPopularity[$offset1];
                            $pop2 = $this->objectsPopularity[$offset2];
                            if ($pop1 == $pop2) {
                                return 0;
                            }
                            if ($pop1 < $pop2) {
                                return -1;
                            }

                            return 1;
                        }
                    );
                    $num = ceil(
                        count($this->collectionGcHold)
                        * $this->config['purgePressure'] / 100
                    );
                    reset($this->collectionGcHold);
                    for ($i = 0; $i < $num; $i++) {
                        $this->releaseHold(key($this->collectionGcHold));
                    }
                    break;
                case Constants::PURGE_STRATEGY_RANDOM:
                    $offsets = array_keys($this->collectionGcHold);
                    shuffle($offsets);
                    $num = ceil (
                        count($this->collectionGcHold)
                        * $this->config['purgePressure'] / 100
                    );
                    $offset = reset($offsets);
                    for ($i = 0; $i < $num; $i++) {
                        $this->releaseHold($offset);
                        $offset = next($offsets);
                    }
                    break;
            }
        }

        /**
         * Replace config options with new values specified in the $config
         * array. For a description of the available config options, see
         * {@see __construct()}
         *
         * @param array $config
         * @throws Exception
         */
        public function setConfig(array $config): void {
            $this->config = array_merge(
                $this->config,
                $config
            );
            switch ($this->config['trigger']) {
                case 'maxRetainedObjects':
                    $this->trigger = Constants::TRIGGER_RETAINED_OBJECTS;
                    break;
                case 'maxScriptMemory':
                    $this->trigger = Constants::TRIGGER_SCRIPT_MEMORY;
                    break;
                case 'none':
                    $this->trigger = Constants::TRIGGER_NONE;
                    break;
                default:
                    throw new Exception('Unknown gc trigger');
            }
            switch ($this->config['purgeStrategy']) {
                case 'popularity':
                    $this->purgeStrategy = Constants::PURGE_STRATEGY_POPULARITY;
                    if (!($this->config['popularityDecay'] ?? null)) {
                        $this->config['popularityDecay'] =
                            $this->getPopularityDecay();
                    }
                    break;
                case 'random':
                    $this->purgeStrategy = Constants::PURGE_STRATEGY_RANDOM;
                    break;
                default:
                    throw new Exception('Unknown purge strategy');
            }
        }

        /**
         * @param bool|float|int|string $offset
         */
        protected function increasePopularity($offset): void {
            if (!array_key_exists($offset, $this->objectsPopularity)) {
                $this->objectsPopularity[$offset] = 0;
                $this->objectsPopularityDecayUpdated[$offset] =
                    $this->objectsPopularityDecayCounter;
            }
            $this->objectsPopularity[$offset]++;
            $this->objectsPopularityDecayUpdated[$offset]++;
            $this->objectsPopularityDecayCounter++;
        }

        /**
         * @param bool|float|int|string $offset
         */
        protected function decay($offset): void {
            $decaySpan =
                $this->objectsPopularityDecayCounter
                - $this->objectsPopularityDecayUpdated[$offset];
            $this->objectsPopularity[$offset] =
                $this->objectsPopularity[$offset] *
                exp(
                    $decaySpan
                    * log(1 - $this->config['popularityDecay'])
                );
            $this->objectsPopularityDecayUpdated[$offset] =
                $this->objectsPopularityDecayCounter;
        }

        /**
         * Gets the configuration options currently in effect.
         * @see __construct()
         *
         * @return array
         */
        public function getConfig(): array {
            return $this->config;
        }

        /**
         * When the 'purgeStrategy' config option is 'popularity', a popularity
         * index will be kept for the cached objects. An object's popularity
         * increases by 1 whenever it is accessed from the cache.<br>
         * <br>
         * In order to not keep forever objects that were once very popular but
         * have not been used in a long time, an object's popularity also decays
         * exponentially every time another object is accessed from the cache.
         * <br>
         * <br>
         * The value of the decay parameter can be specified by the
         * 'popularityDecay' config option. This is a very sensitive float value
         * that can be computed using this function.<br>
         * <br>
         * The meaning of the parameters is:<br>
         * <br>
         * If an object has at the present moment a popularity of
         * $initialPopularity, then other objects are accessed from the cache a
         * number of $rounds times, the popularity of the initial object will
         * drop to $targetPopularity. By tweaking the value of
         * 'popularityDecay', you can balance between keeping in the cache more
         * of formerly popular objects or more of recent objects.
         *
         * @see __construct()
         *
         * @param float $initialPopularity
         * @param int $rounds
         * @param float $targetPopularity
         * @return float
         */
        public static function getPopularityDecay(
            float $initialPopularity = 1000,
            int $rounds = 10000,
            float $targetPopularity = 2
        ): float {
            if($targetPopularity <= 1){
                throw new InvalidArgumentException('targetPopularity cannot be <= 1');
            }
            if($targetPopularity > $initialPopularity){
                throw new InvalidArgumentException('targetPopularity cannot be > initialPopularity');
            }
            if($rounds <= 0){
                throw new InvalidArgumentException('rounds must be > 0');
            }
            if($targetPopularity == $initialPopularity){
                return 0;
            }
            $dLow = 0;
            $dHigh = 1;
            $d = ($dLow + $dHigh) / 2;
            do {
                $lastD = $d;
                $pop =
                    $initialPopularity
                    * exp(
                        $rounds
                        * log(1 - $d)
                    );
                if ($pop < $targetPopularity) {
                    $dHigh = $d;
                    $d = ($dLow + $d) / 2;
                } else {
                    $dLow = $d;
                    $d = ($dHigh + $d) / 2;
                }
            } while (
                $d != $lastD &&
                $d != $dLow &&
                $d != $dHigh
            );

            return $d;
        }
    }